File: 0096-unique-binary-search-trees/0096-unique-binary-search-trees.java
class Solution {
    Map<String, Integer> map = new HashMap();
    public int numTrees(int n) {
        return helper(1, n);
    }
    int helper(int start, int end){
        if(start>end){
            return 1;
        }
        int count = 0;
        if(map.containsKey(start+","+end))
            return map.get(start+","+end);
        for(int i=start; i<=end; i++){
            int lefttree = helper(start, i-1);
            int righttree = helper(i+1, end);
            count +=lefttree * righttree;
        }
        map.put(start+","+end, count);
        return count;
    }
}

File: 0096-unique-binary-search-trees/0096-unique-binary-search-trees.py
class Solution:
    def numTrees(self, n: int) -> int:
        memory = {}
        
        def helper(start, end) -> int:
            count = 0
            if start>end:
                return 1
            if (start, end) in memory:
                return memory[(start,end)]
            
            for i in range(start, end+1):
                l = helper(start, i-1)
                r = helper(i+1, end)
                count += l*r
            
            memory[start, end] = count
            return count

        return helper(1, n)
        
        